text/0032-upgrade-bootstrapper.md

```diff
@@ -0,0 +1,152 @@
+   gets installed to the same location as the current equivalent files.  The location of the `config.toml` can be
+   retrieved from the known location used by the `upgrade-bootstrapper` at step 2.  An upgrade file for chainspec
+   modifications should also be installed and its path included in the new `config-next.toml`
+5. when the current upgrade activation point is reached and the node has finished all its work in that era, it exits
+   normally with an exit code 0
+6. the `upgrade-bootstrapper` renames
+     * `casper-node-next` to `casper-node` (replaces the old binary)
+     * `config.toml` to `config-old.toml`
+     * `config-next.toml` to `config.toml`
+7. the `upgrade-bootstrapper` runs the new `casper-node` in migrate-data mode (see below for further details), passing
+   it the path to `config.toml` and `config-old.toml` as args.  The node performs any migration of stored data and
+   config options required, then exits normally with an exit code 0
+8. repeat from step 3
+
+Should the child process stop with a non-zero exit code at any stage, the `upgrade-bootstrapper` should itself panic.
+It is up to the user to restart the `upgrade-bootstrapper` manually.
+
+It is assumed that the `casper-node` binary will live at a known location and that installing an upgraded binary will
+place the binary and its config file in a known location.  If we wish to be more flexible about this (particularly the
+location of the `casper-node` binary) we can handle these locations in a similar way to the location of the
+`casper-node`'s `config.toml`.
+
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### Changes to existing `casper-node` (running in validator mode)
+
+We require the following changes to `casper-node`:
+
+* running in validator mode, when we reach the upgrade activation point, exit gracefully
+* running in validator mode, if the node starts, joins and finds that it has already reached the upgrade activation
+  point, exit gracefully
+* upgrade activation points should be defined in terms of era IDs, not block heights
+* config.toml updated to take the path to the chainspec upgrade file, similar to the `node.chainspec_config_path`
+  option.  Alternatively, just remove this option, and require the chainspec, its upgrade file, accounts.csv and the
+  config.toml file all to live in the same folder
+
+### Data migration
+
+We require to add a new `migrate-data` subcommand to `casper-node`.  We could provide a separate binary to be used
+exclusively for data and config migration, but it seems simpler to just make this a subcommand.  Simpler for the
+installer, for the `upgrade-bootstrapper` and simpler to code.
+
+Migration of any of the databases needs to take the following approach by iterating each value in the DB and doing:
+
+1. parse as the old type
+2. convert it to the new type
```

Comment:
Since the new type is only part of the new binary, this has to be done using the `casper-node-new`. Will `casper-node-new` binary be able to access the LMDB files of the `casper-node` (old) while it's still running?

Reply:
I believe it should be safe to do that, although I describe that as a fallback option, only required if we decide data migration is going to take too long to be performed after the old binary stops.

If we take that approach, these steps will need adjusted so that we don't delete the old data while the old node binary is still running.